Detections/AzureDevOps/AzDOHistoricServiceConnectionAdds.yaml

```diff
@@ -0,0 +1,50 @@
+description: |
+  'This detection builds a allowlist of historic service connection use by Builds and Releases and compares to recent history, flagging growth of service connection use in non manually allowlisted, non historically allowlisted Build/Release runs.
+  This is to determine if someone is hijacking a build/release and adding many service connections in order to abuse or dump credentials from service connections.'
+severity: High
+queryFrequency: 2h
+queryPeriod: 14d
+triggerOperator: gt
+triggerThreshold: 0
+tactics:
+  - Execution
+  - Impact
+  - PrivilegeEscalation
+relevantTechniques:
+  - T1098
+  - T1527
+  - T1003
+query: |
+
+  let timeframe = 14d;
+  let recentTimeframe = 6h;
+  // Ignore Build/Releases with less/equal this number
+  let ServiceConnectionThreshold = 3;
+  // New Connections need to exhibit execution of more "new" connections than this number.
+  let NewConnectionThreshold = 1;
+  // List of Builds/Releases to ignore
+  let BypassDefIds = datatable(DefId:string, Type:string, ProjectName:string)
+  [
+  //"103", "Release", "ProjectA",
+  //"42", "Release", "ProjectB",
+  //"122", "Build", "ProjectB"
+  ];
+  let HistoricDefs = AzureDevOpsAuditing
+  | where TimeGenerated > ago(timeframe) and TimeGenerated < ago(recentTimeframe)
+  | where OperationName == "Library.ServiceConnectionExecuted" 
+  | extend DefId = tostring(Data.DefinitionId), Type = tostring(Data.PlanType), ConnectionId = tostring(Data.ConnectionId)
+  | summarize HistoricCount = dcount(tostring(ConnectionId)), ConnectionNames = make_set(tostring(Data.ConnectionName)) by tostring(DefId), tostring(Type), ProjectId, ProjectName;
+  AzureDevOpsAuditing
+  | where TimeGenerated >= ago(recentTimeframe)
+  | where OperationName == "Library.ServiceConnectionExecuted" 
+  | extend DefId = tostring(Data.DefinitionId), Type = tostring(Data.PlanType), ConnectionId = tostring(Data.ConnectionId)
+  | parse ScopeDisplayName with OrganizationName ' (Organization)'
+  | summarize CurrentCount = dcount(tostring(ConnectionId)), ConnectionNames = make_set(tostring(Data.ConnectionName)) by OrganizationName, tostring(DefId), tostring(Type), ProjectId, ProjectName
+  | where CurrentCount > ServiceConnectionThreshold
+  | join (HistoricDefs) on ProjectId, DefId, Type
+  | join kind=anti BypassDefIds on $left.DefId==$right.DefId and $left.Type == $right.Type and $left.ProjectName == $right.ProjectName
+  | extend link = iif(Type == "Build", strcat('https://dev.azure.com/', OrganizationName, '/', ProjectName, '/_build?definitionId=', DefId) ,strcat('https://dev.azure.com/', OrganizationName, '/', ProjectName, '/_release?_a=releases&view=mine&definitionId=', DefId))
+  | where CurrentCount >= HistoricCount + NewConnectionThreshold
+  | project OrganizationName, ProjectName, DefId, link, RecentDistinctServiceConnections = CurrentCount, HistoricDistinctServiceConnections = HistoricCount, RecentConnections = ConnectionNames, HistoricConnections = ConnectionNames1
```

Comment:
After you have created the StartTime and EndTime values, use an extend at the end to bring StartTime thru - | extend timestamp = StartTime.  this is required for detections.

Reply:
done and checked 
![image](https://user-images.githubusercontent.com/15659043/86049242-a809b500-ba06-11ea-8d78-b43453fe608f.png)